### PR TITLE
Csp 2274 aws config fetcher no ctx creator

### PR DESCRIPTION
**Pull request by ofiriro3**
- Closes https://github.com/elastic/security-team/issues/3384

### CODE REVIEW

File: go.mod
Comment (ofiriro3) [inline go.mod:55]: Once the latest csp changes will be merged - needs to replace it with the new `main` commit SHA/release

File: go.mod
Comment (ofiriro3) [inline go.mod:160]: Why did I remove it?

File: deploy/eks/cloudbeat-ds.yaml
Comment (ofiriro3) [inline deploy/eks/cloudbeat-ds.yaml:1]: Do we want to refine this file as part of this PR ?

Comment (DaveSys911) [reply deploy/eks/cloudbeat-ds.yaml:1]: @ofiriro3 I think it's best to have one yaml.
I'd add the env vars added for eks to the existing cloudbeat yaml file.

File: JUSTFILE
Comment (DaveSys911) [inline JUSTFILE:28]: a few points here:
- as mentioned in the previous comment, I think there shouldn't be a separation between vanilla & eks yaml files so the same thing goes for the just commands.
- I think login-aws & push-ecr can be unified to one command
- U shouldn't use the latest tag to push to ecr, since you'll override what master pushed(I suggest using branch/pr ref for the tag instead).
- I'd unify them and add a [conditional](https://github.com/casey/just#conditional-expressions) to the just recipe . for exp `if env_var("EKS") == "true"` then invoke the EKS relevant step(publish-ecr) instead of kind-load-image.

**Review by DaveSys911 — COMMENTED** [review]
Added 2 comments to address

File: deploy/eks/cloudbeat-ds.yaml
Comment (ofiriro3) [reply deploy/eks/cloudbeat-ds.yaml:1]: I find it much easier to have a single YAML for each cloud provider. 

If you will pay close attention you would see that there are some differences. 
The fetchers that are being spawned are different as for some of the yaml file currently commented out because AWS won't allow you to do certain operations (like cluster-role that is being managed by AWS)

**Review by ofiriro3 — COMMENTED** [review]
Ready for review @amirbenun 

If you have better names for the interfaces please share it with me

File: resources/providers/kubernetes_provider_mock.go
Comment (amirbenun) [inline resources/providers/kubernetes_provider_mock.go:1]: I didn't realize it uses `codegen` underlying.
We should understand what are the best practices for triggering the code generation in our SDLC.
Meaning, do we want to automate it or keep it a manual process?

File: resources/providers/aws/elb_provider.go
Comment (amirbenun) [inline resources/providers/aws/elb_provider.go:18]: A general note: `aws` package name is confusing, maybe `awslib` instead?

File: resources/fetchers/ecr_factory.go
Comment (amirbenun) [inline resources/fetchers/ecr_factory.go:24]: General note, I think we should try pushing all of these preparation lines of code into the `Create` method.
I realized it may mess up the tests though it makes more sense.
Since the test is on the same package as the factory, you can practically do whatever you want from within the test.

**Review by amirbenun — CHANGES_REQUESTED** [review]
Nice job!
I left some comments, the general notes should apply to all different factories/fetchers.
I agree the naming is hard, I understand what you tried to do with the naming.
Maybe we can just remove the ecr/elb... prefix from the describers? WDYT?

File: resources/providers/aws/elb_provider.go
Comment (ofiriro3) [reply resources/providers/aws/elb_provider.go:18]: I struggled a lot with finding the best name for the package.

I just followed the beats convention at the end - the beats creates `aws` package for all the `aws` content.

But to be honest, I think that `awslib` is better, I hope that it won't confuse anyone.

File: go.mod
Comment (ofiriro3) [inline go.mod:132]: Do we need agent libs ?

File: resources/fetchers/ecr_factory.go
Comment (ofiriro3) [reply resources/fetchers/ecr_factory.go]: So as we discussed I have made some research.

In general, go guidelines propose:
- No package-level variables
- No func init

Supporting articles 
https://leighmcculloch.com/posts/tool-go-check-no-globals-no-inits/
https://peter.bourgon.org/blog/2017/06/09/theory-of-modern-go.html

When diving deep a little bit deeper:
The only"approved" cases when init is legit are:
1. Registration of a package (Similar to the factory registration) 
2. Setup and initialization of specific package 
3. Variables initialization if cannot be done in initialization expr
4. running one-time computations

https://www.digitalocean.com/community/tutorials/understanding-init-in-go
https://medium.com/golangspec/init-functions-in-go-eac191b3860a
https://www.developer.com/languages/inti-function-golang/


Factory - 
```go 
func init() {
	Register("postgres", NewPostgreSQLDataStore)
	Register("memory", NewMemoryDataStore)
}
```
https://matthewbrown.io/2016/01/23/factory-pattern-in-golang

Comment (ofiriro3) [reply resources/fetchers/ecr_factory.go]: So After my research, I have been wondering about few things:

1. What are the benefits of having `init()` function in each and every factory?
I guess that the obvious answer would be - because then I can customize each init (for each factory).
But then the question, why would we want to customize each factory?

We could for example - create only one init() function in the abstract factory and register all of the factories together. 
There is no mutual agreement on the right usage of the init(), but in general, the "godfathers" of the language suggest using it just in only extreme cases.

Implementing it this way can help us achieve more readability and reduce the usage of "magic" in our code.
BTW - we don't need to use init() at all and we can just put all of the register functions in the factory CTOR.
(Better explicit than implicit - Zen of python).

2. Should we put more logic in the init function?

So this is a very tricky question if looking at the first bullet - probably not because we shouldn't use init at all - or at least reduce that magic it does.

Implementation-wise, we will need to move the current init functionality to the "CreateForm" function.
But the thing that bothers me here - "CreateFrom" should be a private function. It feels a little bit smelly to test a private function (or at least a function that should be made private).

On the other hand, keeping the current solution where init() creates the factory is extending the "black magic" in our code, and will probably result in less readability.

Comment (ofiriro3) [reply resources/fetchers/ecr_factory.go]: @amirbenun WDT?

File: resources/providers/kubernetes_provider_mock.go
Comment (ofiriro3) [reply resources/providers/kubernetes_provider_mock.go:1]: I added new code guidelines in the Readme.md file

File: resources/fetchers/ecr_factory.go
Comment (ofiriro3) [reply resources/fetchers/ecr_factory.go]: As we spoke with @yashtewari - changes were made

File: resources/fetchers/iam_factory.go
Comment (amirbenun) [inline resources/fetchers/iam_factory.go:31]: Add info log on factory creations (all the factories that are part of this the pr)

**Review by DaveSys911 — APPROVED** [review]
Approved please add a Todo in justfile & eks/cloudbeat-ds.yml
"Refactor with https://kustomize.io/ "

File: resources/fetchers/eks_factory.go
Comment (amirbenun) [inline resources/fetchers/eks_factory.go:17]: I don't really like this pattern but let's try to think of a better way to load it on the next iterations

**Review by amirbenun — COMMENTED** [review]
Nice, a few more before we merge

File: README.md
Comment (ofiriro3) [reply README.md:118]: Sorry, I missed it - was removed

File: go.mod
Comment (ofiriro3) [reply go.mod]: Will do in other commit

File: go.mod
Comment (ofiriro3) [reply go.mod]: fixed

**Review by amirbenun — APPROVED** [review]

**Pull request merged**
